Support OBJECT_SCHEMA_NAME() T-SQL function

aadityavermaa opened the pull request:
### Description

This commit provides support for OBJECT_SCHEMA_NAME() T-SQL function that returns the schema name for schema-scoped objects. The implementation is a C-type function for improved performance with respect to a SQL function.

### Issues Resolved
BABEL-733

3_X PR: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1189
### Test Scenarios Covered ###
* **Use case based -** Yes


* **Boundary conditions -** Yes


* **Arbitrary inputs -** Yes 


* **Negative test cases -** Yes


* **Minor version upgrade tests -** Yes 


* **Major version upgrade tests -** Yes


* **Performance tests -** N/A


* **Tooling impact -** N/A


* **Client tests -** N/A



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
BABEL-733: Support OBJECT_SCHEMA_NAME T-SQL function

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
removed redundant header files from functions.c

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-733: support OBJECT_SCHEMA_NAME T-SQL function

- removed range checks for input parameters from function because it is
  already been done
- added test cases for implementation
- added tests to upgrade schedule files

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
fixed spell error

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
updated tests

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
added constraint tests

Signed-off-by: Aditya Verma <adiityav@amazon.com>

surendravishnoi commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1564,3 +1565,72 @@ smallint_radians(PG_FUNCTION_ARGS)
+			database_id != get_dbid_from_physical_schema_name(namespace_name, true))
+				PG_RETURN_NULL();
+
+		schema_name = pstrdup(get_logical_schema_name(namespace_name, true));
```
no need duplicate the string, cstring_to_text makes a copy

aadityavermaa replied:
Updated in latest commit.

surendravishnoi commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1564,3 +1565,72 @@ smallint_radians(PG_FUNCTION_ARGS)
+	/* lookup namespace_oid in pg_class */
+	relid = get_rel_namespace(object_id);
+	if(OidIsValid(relid) && 
+		pg_class_aclcheck(object_id, user_id, ACL_SELECT) == ACLCHECK_OK)
```
why are we only doing aclcheck for relid and also continuing if check fails?

aadityavermaa replied:
We need to check if the user has permission to view the object, otherwise we output NULL.
But there is no need to continue if the check fails since we have found the object already. Updated it in the latest commit.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1130,4 +1130,108 @@ split_object_name(char *name)
+		if(!isnull)
+		{
+			Form_pg_proc proc = (Form_pg_proc) GETSTRUCT(tuple);
+			if(OidIsValid(proc->oid))
```
Indentation

aadityavermaa replied:
Fixed in latest commit.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1130,4 +1130,108 @@ split_object_name(char *name)
+		{
+			Form_pg_proc proc = (Form_pg_proc) GETSTRUCT(tuple);
+			if(OidIsValid(proc->oid))
+			if (pg_proc_aclcheck(object_id, user_id, ACL_EXECUTE) == ACLCHECK_OK)
```
why do we continue if perm check fails? do we expect to find it in other catalog?

aadityavermaa replied:
I have updated the helper functions in latest commit to not check for user permissions, the main function performs the check and returns NULL if the check fails.

surendravishnoi commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1130,4 +1130,108 @@ split_object_name(char *name)
+						namespace_oid = con->connamespace;
+				}
+				else
+					namespace_oid = con->connamespace;
```
what use case is this? do we have test?

aadityavermaa replied:
In postgres it is possible to have constraints that are not on tables, but not in TSQL. I have added the else condition to generalise the utility of the helper function, so that it can be used for other constraints too.

surendravishnoi replied:
we should have test case of each if/else condition. If this use case is for non table constraints, add a test case for the same.

surendravishnoi commented on test/JDBC/input/BABEL-733-vu-verify.mix:
```diff
@@ -0,0 +1,239 @@
+-- VERIFY
+
+CREATE VIEW babel_733_sch1.babel_733_v2 AS 
```
have we validate the result against SQL server?

aadityavermaa replied:
Yes.

surendravishnoi commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1564,3 +1565,72 @@ smallint_radians(PG_FUNCTION_ARGS)
+		/* Lookup namespace_oid in pg_trigger */
+		namespace_oid = tsql_get_trigger_nsp_oid(object_id, user_id);
+	}
+
```
how are we making sure that we have considered all the catalogs we need to look into?

aadityavermaa replied:
The catalogs that need to be considered as same as that of object_id and object_name, the other schema-scoped objects ( e.g. - rule, queue ) are not supported in babelfish.

surendravishnoi commented on contrib/babelfishpg_tsql/runtime/functions.c:
```diff
@@ -1564,3 +1565,72 @@ smallint_radians(PG_FUNCTION_ARGS)
+		namespace_name = get_namespace_name(namespace_oid);
+
+		if (pg_namespace_aclcheck(namespace_oid, user_id, ACL_USAGE) != ACLCHECK_OK ||
+			database_id != get_dbid_from_physical_schema_name(namespace_name, true))
```
do we have test case for this?

aadityavermaa replied:
The lookup of object is to be done in the database with oid passed in arg 2, which if NULL then lookup is done in current database. So if the namespace is found but arg2 != db_id of database in physical schema name, then we should return NULL. I have added a test case of this in the latest commit. I have also added a comment for the same.

Commit:
addressed review comments

- return NULL if namespace_oid is found in a view but user does not have
  permission
- added comments
- addressed other review comments

Signed-off-by: Aditya Verma <adiityav@amazon.com>

surendravishnoi commented on test/JDBC/input/BABEL-733-vu-prepare.mix:
```diff
@@ -0,0 +1,44 @@
+-- tsql
```
these tests should be added to upgrade tests, add them to latest and 14_6/14_7 schedule file

aadityavermaa replied:
Added in latest commit.

surendravishnoi commented on test/JDBC/input/BABEL-733-vu-prepare.mix:
```diff
@@ -0,0 +1,44 @@
+-- tsql
+CREATE DATABASE babel_733_db1;
```
Also add objects with lower/upper/mixed case names

aadityavermaa replied:
The function OBJECT_SCHEMA_NAME takes Oid of object and database as parameters, so the name of the object/database wouldn't make a difference.

Commit:
BABEL-733:

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
resolved rebase conflicts

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
add upgrade tests

Signed-off-by: Aditya Verma <adiityav@amazon.com>

aadityavermaa committed:
Merge branch 'BABEL_2_X_DEV' into BABEL-733

aadityavermaa committed:
fixed incorrect comment

aadityavermaa committed:
Merge branch 'BABEL_2_X_DEV' into BABEL-733

aadityavermaa committed:
Merge branch 'BABEL_2_X_DEV' into BABEL-733
